// search-insights@2.7.0/lib/dist/insights.js
"use strict";
exports.__esModule = true;
var objectAssign_1 = require("./polyfills/objectAssign");
var objectKeys_1 = require("./polyfills/objectKeys");
objectKeys_1["default"]();
objectAssign_1["default"]();
var _sendEvent_1 = require("./_sendEvent");
var init_1 = require("./init");
var _algoliaAgent_1 = require("./_algoliaAgent");
var _getVersion_1 = require("./_getVersion");
var click_1 = require("./click");
var conversion_1 = require("./conversion");
var view_1 = require("./view");
var _tokenUtils_1 = require("./_tokenUtils");
var package_json_1 = require("../package.json");
/**
 *  AlgoliaAnalytics class
 */
var AlgoliaAnalytics = /** @class */ (function () {
    function AlgoliaAnalytics(_a) {
        var requestFn = _a.requestFn;
        this._endpointOrigin = "https://insights.algolia.io";
        this._anonymousUserToken = true;
        this._userHasOptedOut = false;
        this._useCookie = false;
        this._cookieDuration = 6 * _tokenUtils_1.MONTH;
        // user agent
        this._ua = [];
        this.version = package_json_1.version;
        this.sendEvents = _sendEvent_1.makeSendEvents(requestFn).bind(this);
        this.init = init_1.init.bind(this);
        this.addAlgoliaAgent = _algoliaAgent_1.addAlgoliaAgent.bind(this);
        this.setUserToken = _tokenUtils_1.setUserToken.bind(this);
        this.setAnonymousUserToken = _tokenUtils_1.setAnonymousUserToken.bind(this);
        this.getUserToken = _tokenUtils_1.getUserToken.bind(this);
        this.onUserTokenChange = _tokenUtils_1.onUserTokenChange.bind(this);
        this.clickedObjectIDsAfterSearch = click_1.clickedObjectIDsAfterSearch.bind(this);
        this.clickedObjectIDs = click_1.clickedObjectIDs.bind(this);
        this.clickedFilters = click_1.clickedFilters.bind(this);
        this.convertedObjectIDsAfterSearch =
            conversion_1.convertedObjectIDsAfterSearch.bind(this);
        this.convertedObjectIDs = conversion_1.convertedObjectIDs.bind(this);
        this.convertedFilters = conversion_1.convertedFilters.bind(this);
        this.viewedObjectIDs = view_1.viewedObjectIDs.bind(this);
        this.viewedFilters = view_1.viewedFilters.bind(this);
        this.getVersion = _getVersion_1.getVersion.bind(this);
    }
    return AlgoliaAnalytics;
}());
exports["default"] = AlgoliaAnalytics;
